tracing-subscriber/src/filter/env/mod.rs

```diff
@@ -223,11 +223,16 @@ impl<S: Subscriber> Layer<S> for EnvFilter {
 
     fn enabled(&self, metadata: &Metadata<'_>, _: Context<'_, S>) -> bool {
         let level = metadata.level();
+        let target = metadata.target();
         self.scope
             .with(|scope| {
-                for filter in scope.iter() {
-                    if filter >= level {
-                        return true;
+                for (target_filter, level_filter) in scope.iter() {
+                    if level_filter >= level {
+                        if let Some(filter) = target_filter {
+                            return &filter[..] == &target[..];
```

Comment:
I don't think this is correct. It looks like this returns `true` if the target matches, and `false` if it doesn't match. We don't want to return `false` if the target doesn't match --- there might be another filter lower in the stack that _does_ have a target that matches, so we want to continue iterating.

Instead, I think this needs to be 
```suggestion
                            if &filter[..] == &target[..] { 
                                return true; 
                            }
```

Reply:
Also, elsewhere, target matching is prefixed based:
https://github.com/tokio-rs/tracing/blob/3fa215fd4a5c513b472ae7c55d2c85cd2c8e314f/tracing-subscriber/src/filter/env/directive.rs#L153-L156
This allows a filter like "tokio=debug" to match `tokio`, `tokio::executor`, `tokio::net::driver`, and so on.

Can we change this to 
```suggestion
                            if &target[..].starts_with(&filter[..]) { 
                                return true; 
                            }
```

Reply:
Additionally, if you want to do some more refactoring, I think the ideal thing would be to encapsulate the behavior for matching target filters, so that there's one canonical implementation. We could add a `TargetFilter` newtype like this:
```rust
#[derive(Clone, Debug)]
struct TargetFilter(Option<Arc<str>>);

impl TargetFilter {

    fn matches(&self, target: &str) -> bool {
        if let Some(filter) = self.0 {
            target.starts_with(&filter[..])
        } else {
            true
        }
    }

}
```